lib/std/crypto/argon2.zig

```diff
@@ -0,0 +1,1119 @@
+/// this behaves almost exactly like the secret or salt parameters; the ad parameter is folding 
+/// into the value of the hash. However, this parameter is used for different data. The salt 
+/// should be a random string stored alongside your password. The secret should be a random key 
+/// only usable at hashing time. The ad is for any other data.
+pub const Params = struct {
+    const Self = @This();
+
+    t: u32,
+    m: u32,
+    p: u8,
+    secret: ?[]const u8 = null,
+    ad: ?[]const u8 = null,
+
+    /// Baseline parameters for interactive logins using argon2i type
+    pub const interactive_2i = Self.fromLimits(4, 33554432);
+    /// Baseline parameters for .. using argon2i type
+    pub const moderate_2i = Self.fromLimits(6, 134217728);
+    /// Baseline parameters for offline usage using argon2i type
+    pub const sensitive_2i = Self.fromLimits(8, 536870912);
+
+    /// Baseline parameters for interactive logins using argon2id type
+    pub const interactive_2id = Self.fromLimits(2, 67108864);
+    /// Baseline parameters for .. using argon2id type
+    pub const moderate_2id = Self.fromLimits(3, 268435456);
+    /// Baseline parameters for offline usage using argon2id type
+    pub const sensitive_2id = Self.fromLimits(4, 1073741824);
+
+    /// Create parameters from ops and mem limits
+    pub fn fromLimits(ops_limit: u32, mem_limit: usize) Self {
+        const m = mem_limit / 1024;
+        std.assert.debug(m <= max_int);
+        return .{ .t = ops_limit, .m = @intCast(u32, m), .p = 1 };
+    }
+};
+
+fn initHash(
+    password: []const u8,
+    salt: []const u8,
+    params: Params,
+    dk_len: usize,
+    mode: Mode,
+) H0 {
+    var h0: H0 = undefined;
+    var parameters: [24]u8 = undefined;
+    var tmp: [4]u8 = undefined;
+    var b2 = Blake2b512.init(.{});
+    mem.writeIntLittle(u32, parameters[0..4], params.p);
+    mem.writeIntLittle(u32, parameters[4..8], @intCast(u32, dk_len));
+    mem.writeIntLittle(u32, parameters[8..12], params.m);
+    mem.writeIntLittle(u32, parameters[12..16], params.t);
+    mem.writeIntLittle(u32, parameters[16..20], version);
+    mem.writeIntLittle(u32, parameters[20..24], @enumToInt(mode));
+    b2.update(&parameters);
+    mem.writeIntLittle(u32, &tmp, @intCast(u32, password.len));
+    b2.update(&tmp);
+    b2.update(password);
+    mem.writeIntLittle(u32, &tmp, @intCast(u32, salt.len));
+    b2.update(&tmp);
+    b2.update(salt);
+    const secret = params.secret orelse "";
+    std.debug.assert(secret.len <= max_int);
+    mem.writeIntLittle(u32, &tmp, @intCast(u32, secret.len));
+    b2.update(&tmp);
+    b2.update(secret);
+    const ad = params.ad orelse "";
+    std.debug.assert(ad.len <= max_int);
+    mem.writeIntLittle(u32, &tmp, @intCast(u32, ad.len));
+    b2.update(&tmp);
+    b2.update(ad);
+    b2.final(h0[0..Blake2b512.digest_length]);
+    return h0;
+}
+
+fn blake2bHash(out: []u8, in: []const u8, comptime hasher: type) void {
```

Comment:
The function name is confusing, as this is not the BLAKE2b hash function, but a construction based on it that allows for an output larger than the block size.

The reference implementation calls it `blake2b_long`, so maybe we should use something similar to clarify that this is not a copy of the regular BLAKE2b implementation.